### PR TITLE
SDPs Fidelity of Separability

### LINKED ISSUE #141: Feature: Fidelity of separability
The fidelity of separability (also known as the maximum separable fidelity [1, 2]) is an entanglement measure that can be approximated via semidefinite programs outlined in [3]. Ancillary files provided with [3] ([here](https://arxiv.org/src/2303.07911v1/anc)) implement SDPs that approximate the fidelity of separability.  

This task would entail:

1. Adding and generalizing the channel and density matrix SDPs from here and placing them into `state_metrics/fidelity_of_separability.py`. 
2. Adding corresponding unit tests in `tests/test_state_metrics/test_fidelity_of_separability.py`
3. Updating docs in `docs/states.rst` under "Distance Metrics for Quantum States"

[1] [Two-message quantum interactive proofs and the quantum separability problem](https://arxiv.org/abs/1211.6120)
[2] [Quantum interactive proofs and the complexity of separability testing](https://arxiv.org/abs/1308.5788)
[3] [Quantum Steering Algorithm for Estimating Fidelity of Separability](https://arxiv.org/abs/2303.07911)

### PR DESCRIPTION
**Pull request by purva-thakre**
## Description
Fixes #141 

Second try to fix this issue because I can never manage to rebase properly. :sweat_smile:

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] Update docstring with more information about the SDP
  - [x] Clarify these are only for pure states
  - [x] Examples in the docstring
  - [x] Create a new issue & assign to myself for mixed states ([link to the issue](https://github.com/vprusso/toqito/issues/166#issue-1738859267))

## Questions
-  N/A

## Status
-  [x] Ready to go

### CODE REVIEW

**Comment by purva-thakre** [comment]
@vprusso This is ready for a review. 

When I run `make html`, `autodocsummary` was adding some more files. I did not include them in this PR but if you want me to, I can do so.

File: tests/test_state_metrics/test_fidelity_of_separability.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        np.array([[1, 0], [0, 0]]),
+        # System A:
+        np.array([[1, 0], [0, 0]]),
+        # System R:
+        np.array([[0, 0], [0, 1]])
+    )
+
+
+def test_errors_state_SDP():
+    """Tests for riased errors in state SDP function."""
```
Comment (purva-thakre) [inline tests/test_state_metrics/test_fidelity_of_separability.py:38]:
```suggestion
    """Tests for raised errors in state SDP function."""
```

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
+"""Add functions for channel fidelity of Separability as defined in.
+Need to cite paper here.
```
Comment (purva-thakre) [inline toqito/channel_metrics/fidelity_of_separability.py:2]:
```suggestion
"""Add functions for channel fidelity of Separability as defined in [Phil23]_.
```

File: tests/test_channel_metrics/test_fidelity_of_separability.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        # System A:
+        np.array([[1, 0], [0, 0]]),
+        # System R:
+        np.array([[0, 0], [0, 1]])
+    )
+
+mixed_rho = max_mixed(8, is_sparse=False)
+
+def test_errors_channel_SDP():
+    """Tests for riased errors in state SDP function."""
```
Comment (purva-thakre) [inline tests/test_channel_metrics/test_fidelity_of_separability.py:30]:
```suggestion
    """Tests for raised errors in state SDP function."""
```

File: docs/_autosummary/toqito.channel_metrics.diamond_norm.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿toqito.channel\_metrics.diamond\_norm
```
Comment (vprusso) [inline docs/_autosummary/toqito.channel_metrics.diamond_norm.rst:1]: Hmm. I'm not sure if the content in the `_autosummary` should be added in git.

File: tests/test_state_metrics/test_fidelity_of_separability.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import numpy as np
+import pytest
+from toqito.states import bell
+from toqito.state_metrics import fidelity_of_separability
+from toqito.matrix_ops import tensor
+
+
+# test mixed rho
+mixed_rho = 1/2 * bell(0) * bell(0).conj().T + 1 / 2 * bell(
```
Comment (vprusso) [inline tests/test_state_metrics/test_fidelity_of_separability.py:11]: For vector multiplication, I think we should prefer using the `@` operator over the `*` operator.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    If you would like to instead use the benchmark introduced in Appendix H,
+    go to :obj:`toqito.state_metrics.fidelity_of_separability`.
+
+    In [Phil23]_ a variational quantum algorithm (VQA) is introduced to test
+    the separability of a general bipartite state. The algorithm utilizes
+    quantum steering between two separated systems such that the separability
+    of the state is quantified.
+
+    Due to the limitations of currently available quantum computers, two
+    optimization semi-definite programs (SDP) benchmarks were introduced to
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:29]: Nitpick: Your preference, but throughout the rest of `toqito`, the term "semidefinite" is preferred over "semi-definite".

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        "Quantum Steering Algorithm for Estimating Fidelity of Separability"
+        https://arxiv.org/abs/2303.07911
+
+    .. [TBWat18] Watrous, John.
+        “The Theory of Quantum Information”
+        Cambridge University Press, 2018
+
+    Args:
+        psi: the density matrix for the tripartite state of interest psi_{BAR}
+        psi_dims: the dimensions of System A, B & R in
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:116]: Nitpick: comma after "B".

Note: I really should hook up automated linting to save you from having to make all these minor changes :). I'll make myself a note to do that to spare future developers the annoyance of me making these types of comments.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+            the input state density matrix. It is assumed that the first
+            quantity in this list is the dimension of System B.
+        k: value for k-extendibility
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:119]: Nitpick: Period at end of statement.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    Raises:
+        AssertionError:
+            * If the provided dimensions are not for a tripartite density \n
+            matrix
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:123]: Nitpick: period at end of statement.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+    Args:
+        psi: the density matrix for the tripartite state of interest psi_{BAR}
+        psi_dims: the dimensions of System A, B & R in
+            the input state density matrix. It is assumed that the first
+            quantity in this list is the dimension of System B.
+        k: value for k-extendibility
+    Raises:
+        AssertionError:
+            * If the provided dimensions are not for a tripartite density \n
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:122]: Hmm, is this `\n` needed?

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    sys_ext = list(range(2, 2 + k - 1))
+    dim_choi = dim_r * (dim_a**k)
+
+    # Projection onto symmetric subspace on AA'.
+    pi_sym = symmetric_projection(dim_a, 2)
+
+    choi = picos.HermitianVariable("S", (dim_choi, dim_choi))
+    choi_partial = picos.partial_trace(choi, sys_ext, choi_dims)
+    sym_choi = symmetric_projection(dim_a, k)
+    problem = picos.Problem(verbosity=2)
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:161]: We may wish to have the `verbosity` argument as an optional argument to the function with a default value of `2`.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+    problem.set_objective(
+        "max",
+        np.real(picos.trace(
+            pi_sym *
+            picos.partial_trace(
+                (picos.partial_transpose(
+                    psi, [0], psi_dims) @ picos.I(dim_a)) *
+                permute_systems(
+                    choi_partial @ picos.I(dim_b*dim_a), [
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:171]: Nitpick: white space between the `*` operator.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+    problem.add_constraint(
+        picos.partial_trace(choi, list(range(
+            1, k+1)), choi_dims) == picos.I(dim_r)
+    )
+    problem.add_constraint(choi >> 0)
+
+    # k-extendablility of Choi state
+    problem.add_constraint(
+        (picos.I(dim_r) @ sym_choi) * choi * (picos.I(dim_r)@sym_choi) == choi
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:186]: Nitpick: white space between the `@` operator.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    )
+
+    # PPT condition on Choi state
+    sys = []
+    for i in range(1, 1+k):
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:191]: Nitpick: white space between the `+` operator.

File: toqito/channel_metrics/fidelity_of_separability.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+        sys = sys + [i]
+        problem.add_constraint(
+            picos.partial_transpose(choi, sys, choi_dims) >> 0)
+
+    solution = problem.solve(solver="cvxopt")
```
Comment (vprusso) [inline toqito/channel_metrics/fidelity_of_separability.py:196]: As suggested with the `verbosity` argument, I wonder if we want to pass in the solver as a default argument to the function with a default value of `"cvxopt"`.

File: toqito/state_metrics/fidelity_of_separability.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+The constrainsts for this function are positive partial transpose(PPT)
+& k-extendible states.
+"""
+
+from toqito.state_props import is_pure, has_symmetric_extension
+from toqito.matrix_props import is_density
+import picos
+import numpy as np
+from toqito.perms import symmetric_projection
```
Comment (vprusso) [inline toqito/state_metrics/fidelity_of_separability.py:11]: Nitpick: group the `toqito` imports together.

Another note for future me is to leverage the `isort` package to do this automatically.

**Review by vprusso — CHANGES_REQUESTED** [review]
Thank you for the updated diff! Overall it looks great. I added a number of very minor things, but modulo those, I think it's close to being done!

File: toqito/matrix_props/is_orthonormal.py [hunk 17]
```diff
@@ -1,6 +1,6 @@
 """Tests for is_orthonormal."""
 import numpy as np
-from toqito.state_props import is_mutually_orthogonal
+from toqito.state_props.is_mutually_orthogonal import is_mutually_orthogonal
```
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/matrix_props/is_orthonormal.py:3]: There was a build failure related to this. I fixed it by specifying the full path. 

https://github.com/purva-thakre/toqito/actions/runs/5166169714/jobs/9306093530#step:5:902

File: toqito/channel_metrics/fidelity_of_separability.py (hunk 7 above)
Comment (purva-thakre) [reply toqito/channel_metrics/fidelity_of_separability.py:116]: I tried to use black & flake8 for the entire `toqito` repo. There were quite a few reported issues. 

Plus, I was not sure what things you wanted to be highlighted as an error & what was acceptable. For example, there are quite a few lines in this PR itself that are longer than what is allowed by flake8.

File: toqito/channel_metrics/fidelity_of_separability.py (hunk 10 above)
Comment (purva-thakre) [reply toqito/channel_metrics/fidelity_of_separability.py:122]: Yes, if I enforce the lentgh of a line has only 79 characters. But we could disregard these in the docstrings, maybe?

File: toqito/channel_metrics/fidelity_of_separability.py (hunk 7 above)
Comment (vprusso) [reply toqito/channel_metrics/fidelity_of_separability.py:116]: Yeah, `black` out of the box might make certain changes that we don't necessarily want. It's on my list of things to do to have a configuration file that is a bit more sensible to conform to the `toqito` project. Running `black` on the project is an eventual goal, but probably outside of the scope for this issue :)

File: toqito/channel_metrics/fidelity_of_separability.py (hunk 10 above)
Comment (vprusso) [reply toqito/channel_metrics/fidelity_of_separability.py:122]: Oh, I mean that I don't think you need the explicit `\n` character. Yes, you'll need a line break, but (I don't believe) including `\n` is necessary, right?

**Comment by vprusso** [comment]
Responded to the outstanding comments, but of course, do let me know if I can chime in with anything else here that might be helpful!

File: toqito/channel_metrics/fidelity_of_separability.py (hunk 7 above)
Comment (purva-thakre) [reply toqito/channel_metrics/fidelity_of_separability.py]: > It's on my list of things to do to have a configuration file that is a bit more sensible to conform to the toqito project.

Ok. If you want, I can also work on getting to this stage. The first step would be manually making the changes that flake8 complains about. 

I am used to how `mitiq` has [different workflows](https://github.com/unitaryfund/mitiq/tree/master/.github/workflows) to check code style, running tests for the code (which toqito does have now with a recent PR) & another to build the documentation from a PR.

Comment (vprusso) [reply toqito/channel_metrics/fidelity_of_separability.py]: Yes, `mitiq` has quite a bit more checks in place which is great. Any inspiration you'd like to take from the `mitiq` project that could benefit the `toqito` workflow I'm all for!

File: tests/test_state_metrics/test_fidelity_of_separability.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+from toqito.matrix_ops import tensor
+
+
+# test mixed rho
+mixed_rho = 1/2 * bell(0) @ bell(0).conj().T + 1 / 2 * bell(
+    3) @ bell(3).conj().T
+
+# test pure entangled rho
+u = bell(0)
+entangled_rho = u * u.conj().T
```
Comment (vprusso) [inline tests/test_state_metrics/test_fidelity_of_separability.py:17]: Generally, one uses `@` for vector/matrix multiplication and `*` for scalar multiplication.

File: tests/test_state_metrics/test_fidelity_of_separability.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            match="Provided input state is entangled."):
+        fidelity_of_separability(entangled_rho, [2, 2])
+    with pytest.raises(
+            ValueError,
+            match="This function only works for pure states."):
+        fidelity_of_separability(mixed_rho, [2, 2])
+
+
+def test_sdp_output():
+    """Test expected output of both the SDP functions."""
```
Comment (vprusso) [inline tests/test_state_metrics/test_fidelity_of_separability.py:59]: It says "both SDP functions" in the comment, but I only see one being tested here.

File: tests/test_state_metrics/test_fidelity_of_separability.py [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    expected_value = 1
```
Comment (vprusso) [inline tests/test_state_metrics/test_fidelity_of_separability.py:60]: Nitpick: We don't necessarily need to define this to be its own variable.

File: toqito/state_metrics/fidelity_of_separability.py [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+The constrainsts for this function are positive partial transpose (PPT)
+& k-extendible states.
+"""
+
+import picos
+import numpy as np
+
+from toqito.perms import symmetric_projection
+from toqito.state_props import is_pure
+from toqito.state_props.is_separable import is_separable
```
Comment (vprusso) [inline toqito/state_metrics/fidelity_of_separability.py:12]: Think you can get away with just:

```suggestion
from toqito.state_props import is_separable

```

File: toqito/matrix_props/is_orthonormal.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vprusso) [reply toqito/matrix_props/is_orthonormal.py:3]: Hmm, I'm not sure if that build failure is due to the abbreviated import. Seems like it's referring to something else?

**Review by vprusso — COMMENTED** [review]
Overall super minor things, but I think it's nearly ready to go! Next iteration I suspect we'll chuck it in! Thank you for the work, @purva-thakre !

File: toqito/state_metrics/fidelity_of_separability.py (hunk 21 above)
Comment (purva-thakre) [reply toqito/state_metrics/fidelity_of_separability.py:12]: Same as above, there was a build failure issue.

File: toqito/matrix_props/is_orthonormal.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/matrix_props/is_orthonormal.py:3]: Seems to be related to `is_mutually_orthogonal` where I get `TypeError: 'module' object is not callable`

- https://github.com/purva-thakre/toqito/actions/runs/5166169714/jobs/9306093530#step:5:680
- https://github.com/purva-thakre/toqito/actions/runs/5166169714/jobs/9306093530#step:5:711

File: toqito/state_metrics/fidelity_of_separability.py (hunk 21 above)
Comment (purva-thakre) [reply toqito/state_metrics/fidelity_of_separability.py:12]: @vprusso I'll go ahead & accept this. Seems `is_separable` `not callable` issue might have been fixed with the linked StackExchange answer. 

I tested the `is_mutually_orthogonal` without specifying the full path but I got the same errors.

**Comment by vprusso** [comment]
Thanks again for the changes, @purva-thakre. Let me know if this is good to go. I believe it is, but wanted to check with you to be sure.

**Comment by purva-thakre** [comment]
@vprusso It is, except for the build failures related to the abbreviated import.  I am not exactly sure as to what is causing them.

**Comment by vprusso** [comment]
> @vprusso It is, except for the build failures related to the abbreviated import. I am not exactly sure as to what is causing them.

Gotcha. In that case, let's just do the more explicit imports to solve the build failures. Sound like a plan?

**Comment by purva-thakre** [comment]
@vprusso Sounds like a plan! This is ready to be merged then.

**Pull request merged**